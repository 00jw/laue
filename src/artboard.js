export default {
  name: 'LaArtboard',

  props: {
    data: {
      type: Array,
      default: () => []
    },

    height: {
      type: [Number, String],
      default: '25%'
    },

    width: {
      type: [Number, String],
      default: '100%'
    }
  },

  provide() {
    return {
      Artboard: this
    }
  },

  render(h) {
    const {width, height} = this

    return h(
      'svg',
      {
        attrs: {
          width,
          height
        }
      },
      this.$slots.default
    )
  }
}
