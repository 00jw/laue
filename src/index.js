import Artboard from './objects/artboard'
import Line from './charts/line'
import Area from './charts/area'

export function Laue(Vue) {
  Vue.component(Artboard.name, Artboard)
  Vue.component(Line.name, Line)
  Vue.component(Area.name, Area)
}

export {Line, Artboard, Area}
