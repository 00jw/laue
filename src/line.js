import Chart from './mixins/chart'
import {getCoordinates} from './utils/point'
import line from 'd3-shape/src/line'
import cardinal from 'd3-shape/src/curve/cardinal'
import {int} from './utils/math'

export default {
  name: 'LaLine',

  mixins: [Chart],

  props: {
    color: {
      default: 'black',
      type: String
    },

    smooth: Boolean,

    dot: Boolean,

    width: {
      type: Number,
      default: 1
    }
  },

  render(h) {
    const {x0, y0, x, y} = this.canvas
    const points = getCoordinates(this.values, x0, y0, x, y)
    const l = line()

    this.smooth && l.curve(cardinal)

    return h('g', [
      h('path', {
        attrs: {
          stroke: this.color,
          fill: 'none',
          'stroke-width': this.width,
          d: l(points)
        }
      }),
      this.dot &&
        points.map(p =>
          h('circle', {
            attrs: {
              cx: p[0],
              cy: p[1],
              r: int(this.width) + 1,
              stroke: '#fff',
              fill: this.color
            }
          })
        )
    ])
  }
}
