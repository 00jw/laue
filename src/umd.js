import {Laue} from './index'

if (typeof window.Vue !== 'undefined') {
  window.Vue.use(Laue)
}

export * from './index'
