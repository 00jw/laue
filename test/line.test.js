import {artboardMount, DATA} from './util'
import {Line, Artboard} from '../src'
import {curveBasis} from 'd3-shape'

describe('Props', () => {
  it(':prop - render correctly', () => {
    const wrapper = artboardMount(h => [
      h(Line, {
        props: {
          prop: 'pv'
        }
      })
    ])

    expect(wrapper.find('path').attributes().d).toBeTruthy()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it(':prop - empty data', () => {
    const wrapper = artboardMount(h => [h(Line)])

    expect(wrapper.find('path').attributes().d).toBeUndefined()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it(':color - default', () => {
    const wrapper = artboardMount(h => [h(Line)])

    expect(wrapper.find('path').attributes().stroke).toBe(
      Artboard.props.colors.default()[0]
    )
  })

  it(':color - custom color', () => {
    const wrapper = artboardMount(h => [
      h(Line, {
        props: {
          color: 'red'
        }
      })
    ])

    expect(wrapper.find('path').attributes().stroke).toBe('red')
  })

  it(':width', () => {
    const wrapper = artboardMount(h => [
      h(Line, {
        props: {
          width: 30
        }
      })
    ])

    expect(wrapper.find('path').attributes()['stroke-width']).toBe('30')
  })

  it(':dot', () => {
    const wrapper = artboardMount(h => [
      h(Line, {
        props: {
          dot: true,
          prop: 'pv'
        }
      })
    ])

    expect(wrapper.findAll('circle').length).toEqual(DATA.length)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it(':dashed - bool type', () => {
    const wrapper = artboardMount(h => [
      h(Line, {
        props: {
          dashed: true
        }
      })
    ])

    expect(wrapper.find('path').element.style['stroke-dasharray']).toBe('3')
  })

  it(':dashed - string type', () => {
    const wrapper = artboardMount(h => [
      h(Line, {
        props: {
          dashed: '3 4'
        }
      })
    ])

    expect(wrapper.find('path').element.style['stroke-dasharray']).toBe('3 4')
  })

  it(':hideLine', () => {
    const wrapper = artboardMount(h => [
      h(Line, {
        props: {
          hideLine: true,
          prop: 'pv'
        }
      })
    ])

    expect(wrapper.contains('path')).toBeFalsy()
  })

  it(':hideLine - hide the line but render dots', () => {
    const wrapper = artboardMount(h => [
      h(Line, {
        props: {
          hideLine: true,
          dot: true,
          prop: 'pv'
        }
      })
    ])

    expect(wrapper.findAll('circle').length).toBe(DATA.length)
  })

  it(':curve - smoosh line', () => {
    const wrapper = artboardMount(h => [
      h(Line, {
        props: {
          curve: true,
          prop: 'pv'
        }
      })
    ])

    expect(wrapper.html()).toMatchSnapshot()
  })

  it(':curve - custom function', () => {
    const wrapper = artboardMount(h => [
      h(Line, {
        props: {
          curve: curveBasis,
          prop: 'pv'
        }
      })
    ])

    expect(wrapper.html()).toMatchSnapshot()
  })

  it(':continued - false', () => {
    const wrapper = artboardMount(
      h => [
        h(Line, {
          props: {
            continued: false,
            dot: true,
            prop: 'pv'
          }
        })
      ],
      {
        props: {
          data: [{pv: 100}, {pv: 200}, {}, {pv: 300}, {pv: 400}]
        }
      }
    )

    expect(wrapper.findAll('circle').length).toBe(4)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it(':continued - true', () => {
    const wrapper = artboardMount(
      h => [
        h(Line, {
          props: {
            continued: false,
            prop: 'pv'
          }
        })
      ],
      {
        props: {
          data: [{pv: 100}, {pv: 200}, {}, {pv: 300}, {pv: 400}]
        }
      }
    )

    expect(wrapper.html()).toMatchSnapshot()
  })

  it(':showValue', () => {
    const wrapper = artboardMount(h => [
      h(Line, {
        props: {
          showValue: true,
          prop: 'pv'
        }
      })
    ])

    expect(wrapper.findAll('text').length).toBe(DATA.length)
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('Slots', () => {
  it('default(scoped slot)', () => {
    const wrapper = artboardMount(h => [
      h(Line, {
        props: {
          prop: 'pv'
        },
        scopedSlots: {
          default: props => {
            return h(
              'text',
              {
                class: 'custom-text'
              },
              props.value
            )
          }
        }
      })
    ])

    expect(wrapper.findAll('.custom-text').length).toEqual(DATA.length)
  })
})

describe('Mixing', () => {
  it('Multiple lines', () => {
    const wrapper = artboardMount(h => [
      h(Line, {
        props: {
          prop: 'pv'
        }
      }),
      h(Line, {
        props: {
          prop: 'uv'
        }
      }),
      h(Line, {
        props: {
          prop: 'error-prop'
        }
      })
    ])

    expect(wrapper.findAll('path[d]').length).toBe(2)
    expect(wrapper.html()).toMatchSnapshot()
  })
})
