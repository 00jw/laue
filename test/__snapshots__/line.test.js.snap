// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mixing Multiple lines 1`] = `
<div style="position: relative; width: 600px;">
  <svg width="600" height="300" viewBox="0 0 600 300">
    <g>
      <path stroke="#3fb1e3" fill="none" stroke-width="1" d="M8,292L105.33333333333333,280.12363636363636L202.66666666666666,8L300,198.1509090909091L397.3333333333333,169.36363636363637L494.66666666666663,201.63636363636363L592,185.5" style="stroke-dasharray: none;"></path>
    </g>
    <g>
      <path stroke="#6be6c1" fill="none" stroke-width="1" d="M8,195.1818181818182L105.33333333333333,227.45454545454544L202.66666666666666,259.72727272727275L300,234.55454545454546L397.3333333333333,263.2772727272727L494.66666666666663,247.14090909090908L592,211.64090909090908"
        style="stroke-dasharray: none;"></path>
    </g>
    <g>
      <path stroke="#626c91" fill="none" stroke-width="1" style="stroke-dasharray: none;"></path>
    </g>
  </svg>
</div>
`;

exports[`Props :continued - false 1`] = `
<div style="position: relative; width: 600px;">
  <svg width="600" height="300" viewBox="0 0 600 300">
    <g>
      <path stroke="#3fb1e3" fill="none" stroke-width="1" d="M8,292L154,197.33333333333331M446,102.66666666666666L592,8" style="stroke-dasharray: none;"></path>
      <g stroke="#fff" fill="#3fb1e3">
        <circle cx="8" cy="292" r="2"></circle>
        <circle cx="154" cy="197.33333333333331" r="2"></circle>
        <circle cx="446" cy="102.66666666666666" r="2"></circle>
        <circle cx="592" cy="8" r="2"></circle>
      </g>
    </g>
  </svg>
</div>
`;

exports[`Props :continued - true 1`] = `
<div style="position: relative; width: 600px;">
  <svg width="600" height="300" viewBox="0 0 600 300">
    <g>
      <path stroke="#3fb1e3" fill="none" stroke-width="1" d="M8,292L154,197.33333333333331M446,102.66666666666666L592,8" style="stroke-dasharray: none;"></path>
    </g>
  </svg>
</div>
`;

exports[`Props :curve - custom function 1`] = `
<div style="position: relative; width: 600px;">
  <svg width="600" height="300" viewBox="0 0 600 300">
    <g>
      <path stroke="#3fb1e3" fill="none" stroke-width="1" d="M8,292L24.222222222222218,290.02060606060604C40.44444444444444,288.04121212121214,72.88888888888889,284.0824242424242,105.33333333333333,236.7490909090909C137.77777777777777,189.41575757575757,170.2222222222222,98.70787878787878,202.66666666666666,85.04575757575758C235.1111111111111,71.38363636363637,267.55555555555554,134.76727272727274,300,161.66121212121212C332.4444444444444,188.55515151515155,364.88888888888886,178.95939393939395,397.3333333333333,179.54030303030302C429.77777777777777,180.12121212121212,462.2222222222222,190.87878787878788,494.6666666666667,193.56818181818184C527.1111111111111,196.25757575757575,559.5555555555555,190.87878787878788,575.7777777777777,188.1893939393939L592,185.5"
        style="stroke-dasharray: none;"></path>
    </g>
  </svg>
</div>
`;

exports[`Props :curve - smoosh line 1`] = `
<div style="position: relative; width: 600px;">
  <svg width="600" height="300" viewBox="0 0 600 300">
    <g>
      <path stroke="#3fb1e3" fill="none" stroke-width="1" d="M8,292C8,292,72.88888888888889,327.4569696969697,105.33333333333333,280.12363636363636C137.77777777777777,232.79030303030305,170.22222222222223,21.66212121212121,202.66666666666666,8C235.1111111111111,-5.66212121212121,267.55555555555554,171.2569696969697,300,198.1509090909091C332.44444444444446,225.0448484848485,364.88888888888886,168.7827272727273,397.3333333333333,169.36363636363637C429.77777777777777,169.94454545454545,462.2222222222222,198.9469696969697,494.66666666666663,201.63636363636363C527.1111111111111,204.32575757575756,592,185.5,592,185.5"
        style="stroke-dasharray: none;"></path>
    </g>
  </svg>
</div>
`;

exports[`Props :dot 1`] = `
<div style="position: relative; width: 600px;">
  <svg width="600" height="300" viewBox="0 0 600 300">
    <g>
      <path stroke="#3fb1e3" fill="none" stroke-width="1" d="M8,292L105.33333333333333,280.12363636363636L202.66666666666666,8L300,198.1509090909091L397.3333333333333,169.36363636363637L494.66666666666663,201.63636363636363L592,185.5" style="stroke-dasharray: none;"></path>
      <g stroke="#fff" fill="#3fb1e3">
        <circle cx="8" cy="292" r="2"></circle>
        <circle cx="105.33333333333333" cy="280.12363636363636" r="2"></circle>
        <circle cx="202.66666666666666" cy="8" r="2"></circle>
        <circle cx="300" cy="198.1509090909091" r="2"></circle>
        <circle cx="397.3333333333333" cy="169.36363636363637" r="2"></circle>
        <circle cx="494.66666666666663" cy="201.63636363636363" r="2"></circle>
        <circle cx="592" cy="185.5" r="2"></circle>
      </g>
    </g>
  </svg>
</div>
`;

exports[`Props :prop - empty data 1`] = `
<div style="position: relative; width: 600px;">
  <svg width="600" height="300" viewBox="0 0 600 300">
    <g>
      <path stroke="#3fb1e3" fill="none" stroke-width="1" style="stroke-dasharray: none;"></path>
    </g>
  </svg>
</div>
`;

exports[`Props :prop - render correctly 1`] = `
<div style="position: relative; width: 600px;">
  <svg width="600" height="300" viewBox="0 0 600 300">
    <g>
      <path stroke="#3fb1e3" fill="none" stroke-width="1" d="M8,292L105.33333333333333,280.12363636363636L202.66666666666666,8L300,198.1509090909091L397.3333333333333,169.36363636363637L494.66666666666663,201.63636363636363L592,185.5" style="stroke-dasharray: none;"></path>
    </g>
  </svg>
</div>
`;

exports[`Props :showValue 1`] = `
<div style="position: relative; width: 600px;">
  <svg width="600" height="300" viewBox="0 0 600 300">
    <g>
      <path stroke="#3fb1e3" fill="none" stroke-width="1" d="M8,292L105.33333333333333,280.12363636363636L202.66666666666666,8L300,198.1509090909091L397.3333333333333,169.36363636363637L494.66666666666663,201.63636363636363L592,185.5" style="stroke-dasharray: none;"></path>
      <g fill="#3fb1e3">
        <text x="8" y="292" dy="-0.31em" text-anchor="middle">1000</text>
        <text x="105.33333333333333" y="280.12363636363636" dy="-0.31em" text-anchor="middle">1368</text>
        <text x="202.66666666666666" y="8" dy="-0.31em" text-anchor="middle">9800</text>
        <text x="300" y="198.1509090909091" dy="-0.31em" text-anchor="middle">3908</text>
        <text x="397.3333333333333" y="169.36363636363637" dy="-0.31em" text-anchor="middle">4800</text>
        <text x="494.66666666666663" y="201.63636363636363" dy="-0.31em" text-anchor="middle">3800</text>
        <text x="592" y="185.5" dy="-0.31em" text-anchor="middle">4300</text>
      </g>
    </g>
  </svg>
</div>
`;
